cpp/include/raft/sparse/op/reduce.cuh

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2019-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cusparse_v2.h>
+
+#include <raft/cudart_utils.h>
+#include <raft/sparse/cusparse_wrappers.h>
+#include <raft/cuda_utils.cuh>
+#include <raft/mr/device/allocator.hpp>
+#include <raft/mr/device/buffer.hpp>
+
+#include <raft/sparse/op/sort.h>
+#include <thrust/device_ptr.h>
+#include <thrust/scan.h>
+#include <raft/device_atomics.cuh>
+
+#include <cuda_runtime.h>
+#include <stdio.h>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <algorithm>
+#include <iostream>
+
+#include <raft/sparse/utils.h>
+#include <raft/sparse/convert/csr.cuh>
+#include <raft/sparse/coo.cuh>
+
+namespace raft {
+namespace sparse {
+namespace op {
+
+template <typename value_idx>
+__global__ void compute_duplicates_diffs_kernel(const value_idx *rows,
+                                                const value_idx *cols,
+                                                value_idx *diff, size_t nnz) {
+  size_t tid = blockDim.x * blockIdx.x + threadIdx.x;
+  if (tid >= nnz) return;
+
+  value_idx d = 1;
+  if (tid == 0 || (rows[tid - 1] == rows[tid] && cols[tid - 1] == cols[tid]))
```

Comment:
small nitpick because the formatting is a little ugly: braces?

Reply:
Also is the safeguard here always valid that the second condition will only be accessed when `tid >= 1`? Theory suggests it should but just confirming

Reply:
Thirdly, why is the first entry marked as duplicate?

Reply:
As discussed offline, create #185